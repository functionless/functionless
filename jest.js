const swcJest = require("@swc/jest");
const { config } = require("./swc-config");

function createTransformer() {
  return swcJest.createTransformer(config);
}

module.exports = { createTransformer };
