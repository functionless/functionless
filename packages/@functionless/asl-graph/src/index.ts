export * from "./asl-graph";
export * from "./asl-integration";
export * from "./asl";
export * from "./eval-expr-context";
export * from "./guards";
export * from "./states";
export * from "./step-function-error";
