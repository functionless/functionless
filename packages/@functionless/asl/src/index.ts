export { ASLGraph } from "./asl-graph";
export { ASLIntegration, isASLIntegration } from "./asl-integration";
export { ASL } from "./asl";
export * from "./eval-expr-context";
export * from "./guards";
export {
  Branch,
  Catch,
  Choice,
  CommonFields,
  CommonTaskFields,
  Condition,
  Fail,
  MapTask,
  ParallelTask,
  Parameters,
  Pass,
  Retry,
  State,
  StateMachine,
  States,
  Succeed,
  Task,
  TerminalState,
  Wait,
  isChoiceState,
  isFailState,
  isMapTaskState,
  isParallelTaskState,
  isPassState,
  isState,
  isSucceedState,
  isTaskState,
  isWaitState,
} from "./states";
export { StepFunctionCause, StepFunctionError } from "./step-function-error";
