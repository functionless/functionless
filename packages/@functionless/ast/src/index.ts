export * from "./assert";
export * from "./constants";
export * from "./declaration";
export * from "./ensure";
export * from "./error";
export * from "./expression";
export * from "./find";
export * from "./guards";
export * from "./literal";
export * from "./node-clone";
export * from "./node-ctor";
export * from "./node-kind";
export * from "./node";
export * from "./reflect";
export * from "./resolve-references";
export * from "./s-expression";
export * from "./span";
export * from "./statement";
export * from "./util";
export * from "./visit";
