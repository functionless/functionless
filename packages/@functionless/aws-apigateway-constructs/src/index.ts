export { APIGatewayVTL } from "./api-vtl";
export {
  APIGatewayError,
  APIGatewayRequestOverride,
  ApiBody,
  ApiGatewayContext,
  ApiGatewayInput,
  ApiGatewayResponseOverride,
  ApiMethod,
  ApiMethodKind,
  ApiParameter,
  ApiParameters,
  ApiRequest,
  AwsMethod,
  HttpMethod,
  LambdaMethod,
  LambdaMethodProps,
  MethodProps,
  MockMethod,
  isApiMethodKind,
} from "./api";
