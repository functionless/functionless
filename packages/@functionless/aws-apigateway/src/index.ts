export * from "./api-vtl";
export * from "./api-integration";
