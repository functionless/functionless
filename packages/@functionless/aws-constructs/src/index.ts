export * from "../../aws-appsync-constructs/src/api";
export * from "./appsync";
export * from "./async-synth";
export * from "./aws";
export * from "./event-bridge";
export * from "./event-source";
export * from "./iterable";
export * from "./queue";
export * from "./queue";
export * from "./secret";
export * from "@functionless/aws-stepfunctions-constructs/src/step-function";
export * from "./table";
export * from "./user-pool";
export * from "./util";

// to prevent the closure serializer from trying to import all of functionless.
export const deploymentOnlyModule = true;
