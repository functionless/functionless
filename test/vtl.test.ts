import "jest";
import { $util, AppsyncContext, ResolverFunction } from "../src";
import { reflect } from "../src/reflect";
import { appsyncTestCase } from "./util";

test("empty function returning an argument", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ a: string }>) => {
      return context.arguments.a;
    })
  );
});

test("return literal object with values", () => {
  appsyncTestCase(
    reflect(
      (context: AppsyncContext<{ arg: string; obj: Record<string, any> }>) => {
        const arg = context.arguments.arg;
        const obj = context.arguments.obj;
        return {
          null: null,
          undefined: undefined,
          string: "hello",
          number: 1,
          ["list"]: ["hello"],
          obj: {
            key: "value",
          },
          arg,
          ...obj,
        };
      }
    )
  );
});

test("computed property names", () => {
  appsyncTestCase(
    reflect(
      (context: AppsyncContext<{ arg: string; obj: Record<string, any> }>) => {
        const name = context.arguments.arg;
        const value = name + "_test";
        return {
          [name]: context.arguments.arg,
          [value]: context.arguments.arg,
        };
      }
    )
  );
});

test("null and undefined", () => {
  appsyncTestCase(
    reflect(
      (_context: AppsyncContext<{ arg: string; obj: Record<string, any> }>) => {
        return {
          name: null,
          value: undefined,
        };
      }
    )
  );
});

test("call function and return its value", () => {
  appsyncTestCase(
    reflect(() => {
      return $util.autoId();
    })
  );
});

test("call function, assign to variable and return variable reference", () => {
  appsyncTestCase(
    reflect(() => {
      const id = $util.autoId();
      return id;
    })
  );
});

test("return in-line spread object", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ obj: { key: string } }>) => {
      return {
        id: $util.autoId(),
        ...context.arguments.obj,
      };
    })
  );
});

test("return in-line list literal", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ a: string; b: string }>) => {
      return [context.arguments.a, context.arguments.b];
    })
  );
});

test("return list literal variable", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ a: string; b: string }>) => {
      const list = [context.arguments.a, context.arguments.b];
      return list;
    })
  );
});

test("return list element", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ a: string; b: string }>) => {
      const list = [context.arguments.a, context.arguments.b];
      return list[0];
    })
  );
});

test("push element to array is renamed to add", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      context.arguments.list.push("hello");
      return context.arguments.list;
    })
  );
});

// TODO https://github.com/functionless/functionless/issues/8
// test("push multiple args is expanded to multiple add calls", () => {
//   const template = reflect((context: AppsyncContext<{ list: string[] }>) => {
//     list.push("hello", "world");
//     return list;
//   });

//   const vtl = new VTL();
//   vtl.eval(template.body);
//   const actual = vtl.toVTL();
//   const expected = `$util.qr($context.arguments.list.addAll(['hello']))
//   $util.qr($context.arguments.list.addAll(['world']))
// ${returnExpr("$context.arguments.list")}`;
//   expect(actual).toEqual(expected);
// });

test("if statement", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      if (context.arguments.list.length > 0) {
        return true;
      } else {
        return false;
      }
    })
  );
});

test("return conditional expression", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list.length > 0 ? true : false;
    })
  );
});

test("property assignment of conditional expression", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return {
        prop: context.arguments.list.length > 0 ? true : false,
      };
    })
  );
});

test("for-of loop", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      const newList = [];
      for (const item of context.arguments.list) {
        newList.push(item);
      }
      return newList;
    })
  );
});

test("break from for-loop", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      const newList = [];
      for (const item of context.arguments.list) {
        if (item === "hello") {
          break;
        }
        newList.push(item);
      }
      return newList;
    })
  );
});

test("local variable inside for-of loop is declared as a local variable", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      const newList = [];
      for (const item of context.arguments.list) {
        const i = item;
        newList.push(i);
      }
      return newList;
    })
  );
});

test("for-in loop and element access", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ record: Record<string, any> }>) => {
      const newList = [];
      for (const key in context.arguments.record) {
        newList.push(context.arguments.record[key]);
      }
      return newList;
    })
  );
});

test("template expression", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ a: string }>) => {
      const local = context.arguments.a;
      return `head ${context.arguments.a} ${local}${context.arguments.a}`;
    })
  );
});

test("conditional expression in template expression", () => {
  appsyncTestCase(
    reflect((context: AppsyncContext<{ a: string }>) => {
      return `head ${
        context.arguments.a === "hello" ? "world" : context.arguments.a
      }`;
    })
  );
});

test("map over list", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list.map((item) => {
        return `hello ${item}`;
      });
    })
  ));

test("map over list with in-line return", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list.map((item) => `hello ${item}`);
    })
  ));

test("chain map over list", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list
        .map((item) => `hello ${item}`)
        .map((item) => `hello ${item}`);
    })
  ));

test("chain map over list multiple array", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list
        .map((item, _i, _arr) => `hello ${item}`)
        .map((item, _i, _arr) => `hello ${item}`);
    })
  ));

test("chain map over list complex", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list
        .map((item, i, arr) => {
          const x = i + 1;
          return `hello ${item} ${x} ${arr.length}`;
        })
        .map((item2, ii) => `hello ${item2} ${ii}`);
    })
  ));

test("forEach over list", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list.forEach((item) => {
        $util.error(item);
      });
    })
  ));

test("reduce over list with initial value", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list.reduce((newList: string[], item) => {
        return [...newList, item];
      }, []);
    })
  ));

test("reduce over list without initial value", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list.reduce((str: string, item) => {
        return `${str}${item}`;
      });
    })
  ));

test("map and reduce over list with initial value", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list
        .map((item) => `hello ${item}`)
        .reduce((newList: string[], item) => {
          return [...newList, item];
        }, []);
    })
  ));

test("map and reduce with array over list with initial value", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list
        .map((item) => `hello ${item}`)
        .reduce((newList: string[], item, _i, _arr) => {
          return [...newList, item];
        }, []);
    })
  ));

test("map and reduce and map and reduce over list with initial value", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list
        .map((item) => `hello ${item}`)
        .reduce((newList: string[], item) => {
          return [...newList, item];
        }, [])
        .map((item) => `hello ${item}`)
        .reduce((newList: string[], item) => {
          return [...newList, item];
        }, []);
    })
  ));

test("$util.time.nowISO8601", () =>
  appsyncTestCase(
    reflect(() => {
      return $util.time.nowISO8601();
    })
  ));

test("BinaryExpr and UnaryExpr are evaluated to temporary variables", () =>
  appsyncTestCase(
    reflect(() => {
      return {
        x: -1,
        y: -(1 + 1),
        z: !(true && false),
      };
    })
  ));

test("binary expr in", () => {
  appsyncTestCase(
    reflect<
      ResolverFunction<{ key: string } | { key2: string }, { out: string }, any>
    >(($context) => {
      if ("key" in $context.arguments) {
        return { out: $context.arguments.key };
      }
      return { out: $context.arguments.key2 };
    }),
    {
      executeTemplates: [
        {
          index: 1,
          context: { arguments: { key: "hi" }, source: {} },
          match: { out: "hi" },
        },
        {
          index: 1,
          context: { arguments: { key2: "hello" }, source: {} },
          match: { out: "hello" },
        },
      ],
    }
  );
});

test("binary expr in array", () => {
  appsyncTestCase(
    reflect<ResolverFunction<{ arr: string[] }, { out: string }, any>>(
      ($context) => {
        if (1 in $context.arguments.arr) {
          return { out: $context.arguments.arr[1] };
        }
        return { out: $context.arguments.arr[0] };
      }
    ),
    {
      executeTemplates: [
        {
          index: 1,
          context: { arguments: { arr: ["1", "2"] }, source: {} },
          match: { out: "2" },
        },
        {
          index: 1,
          context: { arguments: { arr: ["1"] }, source: {} },
          match: { out: "1" },
        },
      ],
    }
  );
});

test("binary expr == in if statement", () => {
  appsyncTestCase(
    reflect<ResolverFunction<{ key: string }, { out: string }, any>>(
      ($context) => {
        if ($context.arguments.key == "hello") {
          return { out: "ohh hi" };
        }
        return { out: "wot" };
      }
    ),
    {
      executeTemplates: [
        {
          index: 1,
          context: { arguments: { key: "hello" }, source: {} },
          match: { out: "ohh hi" },
        },
        {
          index: 1,
          context: { arguments: { key: "giddyup" }, source: {} },
          match: { out: "wot" },
        },
      ],
    }
  );
});

test("binary expr =", () => {
  appsyncTestCase(
    reflect<ResolverFunction<{ key: string }, { out: string }, any>>(
      ($context) => {
        if ($context.arguments.key == "help me") {
          $context.arguments.key = "hello";
        }
        if ($context.arguments.key == "hello") {
          return { out: "ohh hi" };
        }
        return { out: "wot" };
      }
    ),
    {
      executeTemplates: [
        {
          index: 1,
          context: { arguments: { key: "hello" }, source: {} },
          match: { out: "ohh hi" },
        },
        {
          index: 1,
          context: { arguments: { key: "giddyup" }, source: {} },
          match: { out: "wot" },
        },
        {
          index: 1,
          context: { arguments: { key: "help me" }, source: {} },
          match: { out: "ohh hi" },
        },
      ],
    }
  );
});

// https://github.com/functionless/functionless/issues/232
test.skip("binary expr +=", () => {
  appsyncTestCase(
    reflect<ResolverFunction<{ key: string }, { out: number }, any>>(
      ($context) => {
        var n = 0;
        if ($context.arguments.key == "hello") {
          n += 1;
        }
        return { out: n };
      }
    ),
    {
      executeTemplates: [
        {
          index: 1,
          context: { arguments: { key: "hello" }, source: {} },
          match: { out: 1 },
        },
        {
          index: 1,
          context: { arguments: { key: "giddyup" }, source: {} },
          match: { out: 0 },
        },
      ],
    }
  );
});
