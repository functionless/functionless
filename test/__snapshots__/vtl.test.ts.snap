// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$util.time.nowISO8601 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#return($util.time.nowISO8601())",
]
`;

exports[`BinaryExpr and UnaryExpr are evaluated to temporary variables 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = {})
#set($v2 = 0 - 1)
$util.qr($v1.put('x', $v2))
#set($v3 = 1 + 1)
#set($v4 = 0 - $v3)
$util.qr($v1.put('y', $v4))
#set($v5 = true && false)
#set($v6 = !$v5)
$util.qr($v1.put('z', $v6))
#return($v1)",
]
`;

exports[`binary expr = 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = $context.arguments.key == 'help me')
#if($v1)
#set($context.arguments.key = 'hello')
$util.qr($context.arguments.key)
#end
#set($v2 = $context.arguments.key == 'hello')
#if($v2)
#set($v3 = {})
$util.qr($v3.put('out', 'ohh hi'))
#set($context.stash.return__val = $v3)
#set($context.stash.return__flag = true)
#return($context.stash.return__val)
#end
#set($v4 = {})
$util.qr($v4.put('out', 'wot'))
#return($v4)",
]
`;

exports[`binary expr = 2`] = `
Object {
  "out": "ohh hi",
}
`;

exports[`binary expr = 3`] = `
Object {
  "out": "wot",
}
`;

exports[`binary expr = 4`] = `
Object {
  "out": "ohh hi",
}
`;

exports[`binary expr == in if statement 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = $context.arguments.key == 'hello')
#if($v1)
#set($v2 = {})
$util.qr($v2.put('out', 'ohh hi'))
#set($context.stash.return__val = $v2)
#set($context.stash.return__flag = true)
#return($context.stash.return__val)
#end
#set($v3 = {})
$util.qr($v3.put('out', 'wot'))
#return($v3)",
]
`;

exports[`binary expr == in if statement 2`] = `
Object {
  "out": "ohh hi",
}
`;

exports[`binary expr == in if statement 3`] = `
Object {
  "out": "wot",
}
`;

exports[`binary expr in 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = !!$context.arguments['key'])
#if($v1)
#set($v2 = {})
$util.qr($v2.put('out', $context.arguments.key))
#set($context.stash.return__val = $v2)
#set($context.stash.return__flag = true)
#return($context.stash.return__val)
#end
#set($v3 = {})
$util.qr($v3.put('out', $context.arguments.key2))
#return($v3)",
]
`;

exports[`binary expr in 2`] = `
Object {
  "out": "hi",
}
`;

exports[`binary expr in 3`] = `
Object {
  "out": "hello",
}
`;

exports[`binary expr in array 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = !!$context.arguments.arr[1])
#if($v1)
#set($v2 = {})
$util.qr($v2.put('out', $context.arguments.arr[1]))
#set($context.stash.return__val = $v2)
#set($context.stash.return__flag = true)
#return($context.stash.return__val)
#end
#set($v3 = {})
$util.qr($v3.put('out', $context.arguments.arr[0]))
#return($v3)",
]
`;

exports[`binary expr in array 2`] = `
Object {
  "out": "2",
}
`;

exports[`binary expr in array 3`] = `
Object {
  "out": "1",
}
`;

exports[`break from for-loop 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($context.stash.newList = [])
#foreach($item in $context.arguments.list)
#set($v1 = $item == 'hello')
#if($v1)
#break
#end
$util.qr($context.stash.newList.addAll([$item]))
#end
#return($context.stash.newList)",
]
`;

exports[`call function and return its value 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#return($util.autoId())",
]
`;

exports[`call function, assign to variable and return variable reference 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($context.stash.id = $util.autoId())
#return($context.stash.id)",
]
`;

exports[`chain map over list 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = [])
#foreach($item in $context.arguments.list)
#set($item = \\"hello \${item}\\")
#set($v2 = \\"hello \${item}\\")
$util.qr($v1.add($v2))
#end
#return($v1)",
]
`;

exports[`chain map over list complex 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = [])
#foreach($item in $context.arguments.list)
#set($i = $foreach.index)
#set($arr = $context.arguments.list)
#set($v2 = $i + 1)
#set($x = $v2)
#set($item2 = \\"hello \${item} \${x} \${arr.length}\\")
#set($ii = $foreach.index)
#set($v3 = \\"hello \${item2} \${ii}\\")
$util.qr($v1.add($v3))
#end
#return($v1)",
]
`;

exports[`chain map over list multiple array 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = [])
#set($v2 = [])
#foreach($item in $context.arguments.list)
#set($_i = $foreach.index)
#set($_arr = $context.arguments.list)
#set($v3 = \\"hello \${item}\\")
$util.qr($v2.add($v3))
#end
#foreach($item in $v2)
#set($_i = $foreach.index)
#set($_arr = $v2)
#set($v4 = \\"hello \${item}\\")
$util.qr($v1.add($v4))
#end
#return($v1)",
]
`;

exports[`computed property names 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($context.stash.name = $context.arguments.arg)
#set($v1 = $context.stash.name + '_test')
#set($context.stash.value = $v1)
#set($v2 = {})
$util.qr($v2.put($context.stash.name, $context.arguments.arg))
$util.qr($v2.put($context.stash.value, $context.arguments.arg))
#return($v2)",
]
`;

exports[`conditional expression in template expression 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v2 = $context.arguments.a == 'hello')
#if($v2)
#set($v1 = 'world')
#else
#set($v1 = $context.arguments.a)
#end
#return(\\"head \${v1}\\")",
]
`;

exports[`empty function returning an argument 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#return($context.arguments.a)",
]
`;

exports[`for-in loop and element access 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($context.stash.newList = [])
#foreach($key in $context.arguments.record.keySet())
$util.qr($context.stash.newList.addAll([$context.arguments.record[$key]]))
#end
#return($context.stash.newList)",
]
`;

exports[`for-of loop 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($context.stash.newList = [])
#foreach($item in $context.arguments.list)
$util.qr($context.stash.newList.addAll([$item]))
#end
#return($context.stash.newList)",
]
`;

exports[`forEach over list 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#foreach($item in $context.arguments.list)
$util.qr($util.error($item))
#end
#return($null)",
]
`;

exports[`if statement 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = $context.arguments.list.length > 0)
#if($v1)
#set($context.stash.return__val = true)
#set($context.stash.return__flag = true)
#return($context.stash.return__val)
#else
#set($context.stash.return__val = false)
#set($context.stash.return__flag = true)
#return($context.stash.return__val)
#end",
]
`;

exports[`local variable inside for-of loop is declared as a local variable 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($context.stash.newList = [])
#foreach($item in $context.arguments.list)
#set($i = $item)
$util.qr($context.stash.newList.addAll([$i]))
#end
#return($context.stash.newList)",
]
`;

exports[`map and reduce and map and reduce over list with initial value 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v2 = [])
#foreach($item in $context.arguments.list)
#set($item = \\"hello \${item}\\")
#set($newList = $v2)
#set($v4 = [])
$util.qr($v4.addAll($newList))
$util.qr($v4.add($item))
#set($v3 = $v4)
#set($v2 = $v3)
#end
#set($v1 = [])
#foreach($item in $v2)
#set($item = \\"hello \${item}\\")
#set($newList = $v1)
#set($v6 = [])
$util.qr($v6.addAll($newList))
$util.qr($v6.add($item))
#set($v5 = $v6)
#set($v1 = $v5)
#end
#return($v1)",
]
`;

exports[`map and reduce over list with initial value 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = [])
#foreach($item in $context.arguments.list)
#set($item = \\"hello \${item}\\")
#set($newList = $v1)
#set($v3 = [])
$util.qr($v3.addAll($newList))
$util.qr($v3.add($item))
#set($v2 = $v3)
#set($v1 = $v2)
#end
#return($v1)",
]
`;

exports[`map and reduce with array over list with initial value 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v2 = [])
#foreach($item in $context.arguments.list)
#set($v3 = \\"hello \${item}\\")
$util.qr($v2.add($v3))
#end
#set($v1 = [])
#foreach($item in $v2)
#set($_i = $foreach.index)
#set($_arr = $v2)
#set($newList = $v1)
#set($v5 = [])
$util.qr($v5.addAll($newList))
$util.qr($v5.add($item))
#set($v4 = $v5)
#set($v1 = $v4)
#end
#return($v1)",
]
`;

exports[`map over list 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = [])
#foreach($item in $context.arguments.list)
#set($v2 = \\"hello \${item}\\")
$util.qr($v1.add($v2))
#end
#return($v1)",
]
`;

exports[`map over list with in-line return 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = [])
#foreach($item in $context.arguments.list)
#set($v2 = \\"hello \${item}\\")
$util.qr($v1.add($v2))
#end
#return($v1)",
]
`;

exports[`null and undefined 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = {})
$util.qr($v1.put('name', $null))
$util.qr($v1.put('value', $null))
#return($v1)",
]
`;

exports[`property assignment of conditional expression 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = {})
#set($v3 = $context.arguments.list.length > 0)
#if($v3)
#set($v2 = true)
#else
#set($v2 = false)
#end
$util.qr($v1.put('prop', $v2))
#return($v1)",
]
`;

exports[`push element to array is renamed to add 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "$util.qr($context.arguments.list.addAll(['hello']))
#return($context.arguments.list)",
]
`;

exports[`reduce over list with initial value 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = [])
#foreach($item in $context.arguments.list)
#set($newList = $v1)
#set($v3 = [])
$util.qr($v3.addAll($newList))
$util.qr($v3.add($item))
#set($v2 = $v3)
#set($v1 = $v2)
#end
#return($v1)",
]
`;

exports[`reduce over list without initial value 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#if($context.arguments.list.isEmpty())
$util.error('Reduce of empty array with no initial value')
#end
#foreach($item in $context.arguments.list)
#if($foreach.index == 0)
#set($v1 = $item)
#else
#set($str = $v1)
#set($v2 = \\"\${str}\${item}\\")
#set($v1 = $v2)
#end
#end
#return($v1)",
]
`;

exports[`return conditional expression 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v2 = $context.arguments.list.length > 0)
#if($v2)
#set($v1 = true)
#else
#set($v1 = false)
#end
#return($v1)",
]
`;

exports[`return in-line list literal 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#return([$context.arguments.a, $context.arguments.b])",
]
`;

exports[`return in-line spread object 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($v1 = {})
$util.qr($v1.put('id', $util.autoId()))
$util.qr($v1.putAll($context.arguments.obj))
#return($v1)",
]
`;

exports[`return list element 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($context.stash.list = [$context.arguments.a, $context.arguments.b])
#return($context.stash.list[0])",
]
`;

exports[`return list literal variable 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($context.stash.list = [$context.arguments.a, $context.arguments.b])
#return($context.stash.list)",
]
`;

exports[`return literal object with values 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($context.stash.arg = $context.arguments.arg)
#set($context.stash.obj = $context.arguments.obj)
#set($v1 = {})
$util.qr($v1.put('null', $null))
$util.qr($v1.put('undefined', $null))
$util.qr($v1.put('string', 'hello'))
$util.qr($v1.put('number', 1))
$util.qr($v1.put('list', ['hello']))
#set($v2 = {})
$util.qr($v2.put('key', 'value'))
$util.qr($v1.put('obj', $v2))
$util.qr($v1.put('arg', $context.stash.arg))
$util.qr($v1.putAll($context.stash.obj))
#return($v1)",
]
`;

exports[`template expression 1`] = `
Array [
  "{
  \\"version\\": \\"2018-05-29\\",
  \\"payload\\": null
}",
  "#set($context.stash.local = $context.arguments.a)
#return(\\"head \${context.arguments.a} \${context.stash.local}\${context.arguments.a}\\")",
]
`;
